benchmark/utils/stencil_matrix.hpp

```diff
@@ -0,0 +1,501 @@
+    auto root_floor = std::floor(root);
+    auto root_ceil = std::ceil(root);
+    if (root - root_floor > root_ceil - root) {
+        return root_ceil;
+    } else {
+        return root_floor;
+    }
+}
+
+
+/**
+ * Generates matrix data for a 2D stencil matrix. If restricted is set to true,
+ * creates a 5-pt stencil, if it is false creates a 9-pt stencil.
+ *
+ * If
+ * strong_scaling is set to true, creates the same problem size independent of
+ * the number of ranks, if it false the problem size grows with the number of
+ * ranks.
+ */
+template <typename ValueType, typename IndexType>
+gko::matrix_data<ValueType, IndexType> generate_2d_stencil_box(
+    std::array<int, 2> dims, std::array<int, 2> position,
+    const gko::size_type target_local_size, bool restricted)
+{
+    auto num_boxes =
+        std::accumulate(dims.begin(), dims.end(), 1, std::multiplies<>{});
+
+    const auto dp =
+        static_cast<IndexType>(closest_nth_root(target_local_size, 2));
+    const auto local_size = static_cast<gko::size_type>(dp * dp);
+    const auto global_size = local_size * num_boxes;
+    auto A_data = gko::matrix_data<ValueType, IndexType>(
+        gko::dim<2>{static_cast<gko::size_type>(global_size),
+                    static_cast<gko::size_type>(global_size)});
+
+    auto global_offset = [&](const int px, const int py) {
+        return static_cast<int>(local_size) * px +
+               static_cast<int>(local_size) * dims[0] * py;
+    };
+
+    auto target_coords = [&](const IndexType i, const int coord) {
+        return 0 <= i && i <= dp - 1 ? coord : i < 0 ? coord - 1 : coord + 1;
+    };
+
+    auto target_local_idx = [&](const IndexType i) {
+        return 0 <= i && i <= dp - 1 ? i : i < 0 ? dp + i : dp - i;
+    };
+
+    auto flat_idx = [&](const IndexType ix, IndexType iy) {
+        return global_offset(target_coords(ix, position[0]),
+                             target_coords(iy, position[1])) +
+               target_local_idx(ix) + target_local_idx(iy) * dp;
+    };
+
+    auto is_valid_idx = [&](const IndexType i) {
+        return i >= 0 && i < static_cast<IndexType>(global_size);
+    };
+
+    auto is_valid_neighbor = [&](const IndexType d_i, const IndexType d_j) {
+        return !restricted || ((d_i == 0 || d_j == 0));
+    };
+
+    auto nnz_in_row = [&]() {
+        int num_neighbors = 0;
+        for (IndexType d_i : {-1, 0, 1}) {
+            for (IndexType d_j : {-1, 0, 1}) {
+                if (is_valid_neighbor(d_i, d_j)) {
+                    num_neighbors++;
+                }
+            }
+        }
+        return num_neighbors;
+    };
+    const auto diag_value = static_cast<ValueType>(nnz_in_row() - 1);
+
+    for (IndexType i = 0; i < dp; ++i) {
+        for (IndexType j = 0; j < dp; ++j) {
+            auto row = flat_idx(j, i);
+            for (IndexType d_i : {-1, 0, 1}) {
+                for (IndexType d_j : {-1, 0, 1}) {
+                    if (is_valid_neighbor(d_i, d_j)) {
```

Comment:
this doesn't seem to do any bounds checking at the boundary, is that intended?

Reply:
This checks only if the point is valid for the given stencil. The actual in-bounds check is in `is_valid_idx`, which is wrong currently, but I have a fix for that. There still seems to be another issue, because in the distributed case it only converges, if the domain is decomposed in 1D. I've yet to find the reason for that.